doc/tutorial.md

```diff
@@ -316,9 +316,7 @@ The function `slow()` deserves more explanation. As usual, this function returns
 This example begins to show the convenience of the futures programming model, which allows the programmer to neatly encapsulate complex asynchronous operations. `slow()` might involve a complex asynchronous operation requiring multiple steps, but its user can use it just as easily as a simple `sleep()`, and Seastar's engine takes care of running the continuations whose futures have become ready at the right time.
 
 ## Ready futures
-A future value might already be ready when `then()` is called to chain a continuation to it. This important case is optimized, and *usually* the continuation is run immediately instead of being registered to run later in the next iteration of the event loop.
-
-This optimization is done *usually*, though sometimes it is avoided: The implementation of `then()` holds a counter of such immediate continuations, and after many continuations have been run immediately without returning to the event loop (currently the limit is 256), the next continuation is deferred to the event loop in any case. This is important because in some cases (such as future loops, discussed later) we could find that each ready continuation spawns a new one, and without this limit we can starve the event loop. It is important not to starve the event loop, as this would starve continuations of futures that weren't ready but have since become ready, and also starve the important **polling** done by the event loop (e.g., checking whether there is new activity on the network card).
+A future value might already be ready when `then()` is called to chain a continuation to it. This important case is optimized, and the continuation is run immediately instead of being registered to run later in the next iteration of the event loop.
```

Comment:
FWIW, the limit was killed in ab74536d0fc312ad56facfe7be54c56f0dceb441 .

Reply:
Heh, in 2016, so better late than never to reflect this in the tutorial :).

Reply:
@travisdowns @tchaikov actually, the commit that @tchaikov found didn't cancel preemption in then()... The change back then was not to do this based on counting (every 256 continuations) but based on the "preemption timer" (need_preempt()). The change you were looking for was f58515a95cbf304ea48e0f5d8d046b7ab4357789 and was done in 2020.

Thanks for looking into this issue, it's indeed very important to document this clearly and correctly, because it's hard to guess what needs to be done to avoid "stalls" (long periods without preemption). We need to explain that then() does **not** do preemptions these days, but our loop functions do (by "doing preemption" I mean only if need_preempt() - which needs to be explained too). And most importantly for modern code based on coroutines: Our "co_await()" **does** (unless I'm misremembering?) check for need_preempt(). Why did we decide in 2020 we need to remove the need_preempt() check from then(), only to bring it back into co_await()? Maybe @avikivity remembers, I don't :-(
I see you started discussing this below. Good.